src/main/java/org/assertj/core/api/AbstractCharSequenceAssert.java

```diff
@@ -1022,6 +1024,29 @@ public SELF matches(CharSequence regex) {
     return myself;
   }
 
+  /**
+   * Verifies that the actual {@code CharSequence} matches the given regular expression pattern.
+   * Accepts a {@code Consumer<Matcher>} to do further verification of the match.
+   * <p>
+   * Example :
+   * <pre><code class='java'> // assertion will pass
+   * assertThat(&quot;Frodo&quot;).matchesSatisfying(&quot;..(o.o)&quot;, match -&gt; assertThat(match).isEqualTo(&quot;odo&quot;));
+   * </code></pre>
+   *
+   * @param regex the regular expression to which the actual {@code CharSequence} is to be matched.
+   * @param matchSatisfies a consumer of the Matcher to do further verification
+   * @return {@code this} assertion object.
+   * @throws NullPointerException if the given pattern is {@code null}.
+   * @throws AssertionError if the actual {@code CharSequence} is {@code null}.
+   * @throws AssertionError if the actual {@code CharSequence} does not match the given regular expression.
+   */
+  public SELF matchesSatisfying(CharSequence regex, Consumer<Matcher> matchSatisfies) {
+    Matcher matcher = Pattern.compile(regex.toString()).matcher(actual);
+    strings.assertMatches(info, actual, matcher);
+    matchSatisfies.accept(matcher);
+    return myself;
```

Comment:
I think this could simply call `matchesSatisfying(Pattern pattern, Consumer<Matcher> matchSatisfies)` as:
```java
return matchesSatisfying(Pattern.compile(regex.toString()), matchSatisfies);
```